### DISCUSSION TITLE
Bottom-up model not working for new mice

### DISCUSSION THREAD

**Discussion by tnspangler0123** [talmolab/sleap#1105] [Help!] [answered]
I trained with 700 frames for this project. SLEAP was working flawlessly with our first pair of mice ("UK mice"). However, when we recently put in the next pair of mice ("YC mice"), the labeling barely works and is entirely inaccurate. I am unable to attach files but will try to add a link: 

The first video with all the user-inputted labels is what I used to train. The second video is a demonstration that the UK mice are labeled and tracked correctly. The third video is the new YC mice which somehow it does not work for. I've tried both greedy and hungarian. Otherwise all the inference settings are the same.

**Comment by talmo** [comment, marked as answer]
Hi @tnspangler0123,

Thanks for providing the troubleshooting data!

This appears to be a classic case of domain shift. This is basically the problem where a model fails to generalize to new data that is sufficiently distinct from the data it was trained on.

Normally, deep neural network models trained on very diverse data can generalize pretty well, but in this case, since the training labels were all from the same video, it's unlikely that the model will perform well on new videos that look different, even in subtle ways.

Here's what's going on with your data. Check out these screenshots from the 3 videos in the project:

UK1 (which your SLEAP model was trained on):
![uk1](https://user-images.githubusercontent.com/3187454/210927665-8f9d27e9-f801-4e7a-979d-303be29bf604.jpg)

UK2 (no labels from this video but the model performs well):
![uk2](https://user-images.githubusercontent.com/3187454/210927676-0aaaa14c-1341-4908-8f83-82d5fcfe2b45.jpg)

And YC (new video, no labels, bad performance):
![yc](https://user-images.githubusercontent.com/3187454/210927738-66b00a92-81e7-43a9-b0f8-a096e9dc0ef2.jpg)

Do you see the difference? The YC video is more desaturated, darker and has a lower dynamic range than the two UK videos. We can see this even more clearly if we plot the distribution of pixel intensities in each:

![image](https://user-images.githubusercontent.com/3187454/210927869-e5d76e08-d92a-4e8f-b4d1-87f4f6802f6b.png)

Note how the image statistics are highly overlapping between the two UK videos but not the YC video.

**Since the SLEAP model only ever saw the image statistics from the UK1 video, it can generalize to UK2 but not YC.**

There are a couple of ways to address this most immediately:

1. Add a few labeled frames (~10-20) from the YC video so the model can adapt to the different image statistics.
2. Try enabling contrast augmentation in the training settings so that SLEAP randomly makes labeled images darker or brighter so that the model learns to be robust to changes in image statistics. This is less preferred since it's less obvious how to adjust the augmentation setting to match the range of contrasts present in your data.

In the future, however, you may instead consider employing a couple of best practices that will avoid this issue:

1. In your camera settings, disable automatic contrast, brightness, black level and exposure. Pick a set of values and keep these fixed so that the camera isn't capturing images that are of varying degrees of brightness. These are usually set to "auto" by default so that the camera adjusts to the illumination conditions by default.
2. Keep your illumination constant. If you're imaging in infrared, this can be achieved by having dedicated illumination LEDs at infrared wavelengths (>800nm) and putting a highpass filter on your camera to only allow the dedicated illumination LED photons to pass into the camera sensor while excluding other sources of light that may be variable like the room light, computer screens or windows.
3. Both for previously collected data that is more variable, as well as new data that follows the above practices, create your SLEAP project with videos sampled from many recording sessions, days, mouse genotypes and experimental conditions. Then, prefer labeling fewer frames per video in favor of the same number of frames but spread out across all the videos. The goal when creating a labeling project is to generate labeled examples that are representative of the full diversity found in your data (e.g., dark images, light images, images collected at different times of day, etc.).

Give these a go and let us know how it goes or if you have any other questions!

Cheers,

Talmo

**Reply by jmdelahanty** [reply]
One thing I'm wondering after reading is:

## How many frames makes up a "representative" example of something?

So far, people I've chatted with use something like 75 frames labeled per video with a couple different poses. It feels like these numbers can be arbitrary since some poses require more examples than others like when a point is hidden by the subject's body for example. Is it possible to over-represent example poses in training data even when you've labeled a ton of frames or is that something to worry about only when you have a small training dataset? Or is it mostly a problem if you over represent image statistics? Or maybe both?

To rephrase it, if I label additional examples of poses that the model seemed to struggle with in a particular video, should I also label the same number of frames in the rest of the videos with different image statistics even if the model doesn't struggle as much with those examples?

## Is it ever valid to try and normalize images in videos if collected videos are slightly different?

If videos have been collected and there's enough of a difference between them, does it ever make sense to try and do some sort of image processing so they are more similar? I'm guessing no since that's a lot of preprocessing that will take a lot of time to do properly and the model can learn so long as you give it more examples, but just another thing that came to mind.
